BLE_GattServer/main.cpp

```diff
@@ -0,0 +1,668 @@
+        printf("\thour characteristic value handle %u\r\n", _hour_char.getValueHandle());
+        printf("\t\ttextual description handle %u\r\n", _hour_char_description.getHandle());
+        printf("\tminute characteristic value handle %u\r\n", _minute_char.getValueHandle());
+        printf("\t\ttextual description handle %u\r\n", _minute_char_description.getHandle());
+        printf("\tsecond characteristic value handle %u\r\n", _second_char.getValueHandle());
+        printf("\t\ttextual description handle %u\r\n", _second_char_description.getHandle());
+
+        _event_queue->call_every(1000 /* ms */, callback(this, &Self::increment_second));
+    }
+
+private:
+
+//
+// Handlers called after a characteristic has been read, writen or notified
+//
+    void when_data_sent(unsigned count)
+    {
+        printf("sent %u updates\r\n", count);
+    }
+
+    void when_data_written(const GattWriteCallbackParams *e)
+    {
+        printf("data written:\r\n");
+        printf("\tconnection handle: %u\r\n", e->connHandle);
+        printf("\tattribute handle: %u", e->handle);
+        if (e->handle == _hour_char.getValueHandle()) {
+            printf(" (hour characteristic)\r\n");
+        } else if (e->handle == _minute_char.getValueHandle()) {
+            printf(" (minute characteristic)\r\n");
+        } else if (e->handle == _second_char.getValueHandle()) {
+            printf(" (second characteristic)\r\n");
+        } else {
+            printf("\r\n");
+        }
+        printf("\twrite operation: %u\r\n", e->writeOp);
+        printf("\toffset: %u\r\n", e->offset);
+        printf("\tlength: %u\r\n", e->len);
+        printf("\t data: ");
+
+        for (size_t i = 0; i < e->len; ++i) {
+            printf("%02X", e->data[i]);
+        }
+
+        printf("\r\n");
+    }
+
+    void when_data_read(const GattReadCallbackParams *e)
+    {
+        printf("data read:\r\n");
+        printf("\tconnection handle: %u\r\n", e->connHandle);
+        printf("\tattribute handle: %u", e->handle);
+        if (e->handle == _hour_char.getValueHandle()) {
+            printf(" (hour characteristic)\r\n");
+        } else if (e->handle == _minute_char.getValueHandle()) {
+            printf(" (minute characteristic)\r\n");
+        } else if (e->handle == _second_char.getValueHandle()) {
+            printf(" (second characteristic)\r\n");
+        } else {
+            printf("\r\n");
+        }
+    }
+
+//
+// Client Characteristic Configuration Descriptors (CCCD) events handlers
+//
+    void when_update_enabled(GattAttribute::Handle_t handle)
+    {
+        printf("update enabled on handle %d\r\n", handle);
+    }
+
+    void when_update_disabled(GattAttribute::Handle_t handle)
+    {
+        printf("update disabled on handle %d\r\n", handle);
+    }
+
+    void when_confirmation_received(GattAttribute::Handle_t handle)
+    {
+        printf("confirmation received on handle %d\r\n", handle);
+    }
+
+//
```

Comment:
these three seem repetitive, maybe characteristic should be a class to avoid all functions having to be in triplicate?